### PR TITLE
Fix OIDC scopes handling in scope validation

### PR DESCRIPTION
**Pull request by shashimalcse**
### Proposed changes in this pull request

$subject

### CODE REVIEW

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/validators/OIDCScopeHandler.java [hunk 1]
```diff
@@ -52,6 +53,7 @@ public boolean validateScope(OAuthTokenReqMessageContext tokReqMsgCtx) throws Id
             // Remove openid scope from the token message context.
             String[] scopes = (String[]) ArrayUtils.removeElement(tokReqMsgCtx.getScope(), OAuthConstants.Scope.OPENID);
             tokReqMsgCtx.setScope(scopes);
```
Comment (mefarazath) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/validators/OIDCScopeHandler.java:55]: This needs to be remove with the new change right?

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/handlers/grant/RefreshGrantHandler.java [hunk 2]
```diff
@@ -149,22 +149,22 @@ public boolean validateScope(OAuthTokenReqMessageContext tokReqMsgCtx)
         String[] requestedScopes = tokReqMsgCtx.getOauth2AccessTokenReqDTO().getScope();
         String[] grantedScopes = tokReqMsgCtx.getScope();
         String[] grantedInternalScopes = tokReqMsgCtx.getAuthorizedInternalScopes();
+        List<String> oidcScopes = tokReqMsgCtx.getOidcScopes();
+        List<String> grantedScopeList = Stream.concat(Arrays.stream(grantedScopes),
```
Comment (mefarazath) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/handlers/grant/RefreshGrantHandler.java:153]: If we are going to check contains wouldnt it be good to use Set

**Review by mefarazath — CHANGES_REQUESTED** [review]

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/handlers/grant/RefreshGrantHandler.java [hunk 3]
```diff
@@ -149,22 +149,22 @@ public boolean validateScope(OAuthTokenReqMessageContext tokReqMsgCtx)
...
-            for (String scope : requestedScopes) {
-                if (!grantedScopeList.contains(scope)) {
-                    if (log.isDebugEnabled()) {
-                        log.debug("scope: " + scope + "is not granted for this refresh token");
-                    }
-                    return false;
-                }
+            List<String> filteredScopes = Arrays.stream(requestedScopes).distinct().filter(grantedScopeList::contains)
+                            .collect(Collectors.toList());
+            if (filteredScopes.isEmpty()) {
```
[the remark this answers is not in the document]
Comment (shashimalcse) [reply components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/handlers/grant/RefreshGrantHandler.java:162]: if requestedScopes is empty we set scope as an empty array in first place right?

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/handlers/grant/RefreshGrantHandler.java [hunk 4]
```diff
@@ -147,28 +149,34 @@ public boolean validateScope(OAuthTokenReqMessageContext tokReqMsgCtx)
...
-            tokReqMsgCtx.setScope(requestedScopes);
+            tokReqMsgCtx.setScope(filteredScopes.toArray(new String[0]));
+        } else {
+            tokReqMsgCtx.setScope(new String[0]);
         }
         return true;
     }
 
+    private HashSet<String> getGrantedScopeList(OAuthTokenReqMessageContext tokReqMsgCtx) {
+        String[] grantedScopes = tokReqMsgCtx.getScope();
```
Comment (shashimalcse) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/handlers/grant/RefreshGrantHandler.java:170]: put a comment

**Pull request closed without merging**